builder/swig/go.bzl

```diff
@@ -0,0 +1,201 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at https://mozilla.org/MPL/2.0/.
+load("@io_bazel_rules_go//go:def.bzl", "go_library")
+
+def _swig_go_wrapper_impl(ctx):
+    module_name = ctx.attr.name
+
+    # Go outputs to a directory.
+    wrap_go_dir = ctx.actions.declare_directory("{}".format(module_name))
+
+    args = ctx.attr.extra_args + [
+        "-go",
+        "-outdir", wrap_go_dir.path,
+        "-package", "typedb_driver",
+        ctx.file.interface.path,
+    ]
+
+    if ctx.attr.enable_cxx:
+        wrap_src = ctx.actions.declare_file("{}_wrap.cxx".format(module_name))
+        directors_header = ctx.actions.declare_file("{}_wrap.h".format(module_name))
+        args = ["-c++", "-o", wrap_src.path, "-oh", directors_header.path] + args
+        swig_headers = [directors_header]
+    else:
+        wrap_src = ctx.actions.declare_file("{}_wrap.c".format(module_name))
+        args = ["-o", wrap_src.path] + args
+        swig_headers = []
+
+    for h in ctx.attr.lib[CcInfo].compilation_context.headers.to_list():
+        args = ["-I" + h.dirname] + args
+
+    ctx.actions.run(
+        inputs = depset(
+            [ctx.file.interface] + ctx.files.includes,
+
+            transitive = [
+                ctx.attr.lib[CcInfo].compilation_context.headers,
+                ctx.attr._swig.data_runfiles.files,
+            ]),
+        outputs = [wrap_src, wrap_go_dir] + swig_headers,
+        executable = ctx.file._swig,
+        arguments = args,
+    )
+
+    lib_compilation_context = ctx.attr.lib[CcInfo].compilation_context
+    compilation_context = cc_common.create_compilation_context(
+        headers = depset(swig_headers, transitive = [lib_compilation_context.headers]),
+        defines = lib_compilation_context.defines,
+        framework_includes = lib_compilation_context.framework_includes,
+        includes = depset(
+            [h.dirname for h in ctx.attr.lib[CcInfo].compilation_context.headers.to_list()],
+            transitive = [lib_compilation_context.includes],
+        ),
+        local_defines = lib_compilation_context.local_defines,
+        quote_includes = lib_compilation_context.quote_includes,
+        system_includes = depset(
+            [],
+            transitive = [lib_compilation_context.system_includes],
+        )
+    )
+
+    # need to use OutputGroupInfo to separate the go file and cxx, so it doesn't attempt to compile both with C and
```

Comment:
We use a similar approach in `csharp` and its usage is quite obvious when you already know Bazel and see how we get use these files later. But I'm not against leaving a comment here if you want to highlight it, just try not to overdo it (code should aim to be [self-documented](https://stackoverflow.com/questions/209015/what-is-self-documenting-code-and-can-it-replace-well-documented-code))